fileManager1API.js

```diff
@@ -0,0 +1,218 @@
+// -*- mode: js; js-indent-level: 4; indent-tabs-mode: nil -*-
+
+const Gio = imports.gi.Gio;
+const Signals = imports.signals;
+
+const Me = imports.misc.extensionUtils.getCurrentExtension();
+const Utils = Me.imports.utils;
+
+const FileManager1Iface = '<node><interface name="org.freedesktop.FileManager1">\
+                               <property name="XUbuntuOpenLocationsXids" type="a{uas}" access="read"/>\
+                               <property name="OpenWindowsWithLocations" type="a{sas}" access="read"/>\
+                           </interface></node>';
+
+const FileManager1Proxy = Gio.DBusProxy.makeProxyWrapper(FileManager1Iface);
+
+/**
+ * This class implements a client for the org.freedesktop.FileManager1 dbus
+ * interface, and specifically for the OpenWindowsWithLocations property
+ * which is published by Nautilus, but is not an official part of the interface.
+ *
+ * The property is a map from window identifiers to a list of locations open in
+ * the window.
+ *
+ * While OpeWindowsWithLocations is part of upstream Nautilus, for many years
+ * prior, Ubuntu patched Nautilus to publish XUbuntuOpenLocationsXids, which is
+ * similar but uses Xids as the window identifiers instead of gtk window paths.
+ *
+ * When an old or unpatched Nautilus is running, we will observe the properties
+ * to always be empty arrays, but there will not be any correctness issues.
+ */
+var FileManager1Client = class DashToDock_FileManager1Client {
+
+    constructor() {
+        this._signalsHandler = new Utils.GlobalSignalsHandler();
+
+        this._locationMap = new Map();
+        this._proxy = new FileManager1Proxy(Gio.DBus.session,
+                                            "org.freedesktop.FileManager1",
+                                            "/org/freedesktop/FileManager1",
+                                            (initable, error) => {
+            // Use async construction to avoid blocking on errors.
+            if (error) {
+                global.log(error);
+            } else {
+                this._updateLocationMap();
+            }
+        });
+
+        this._signalsHandler.add([
+            this._proxy,
+            'g-properties-changed',
+            this._onPropertyChanged.bind(this)
+        ], [
+            // We must additionally listen for Screen events to know when to
+            // rebuild our location map when the set of available windows changes.
+            global.workspace_manager,
+            'workspace-switched',
+            this._updateLocationMap.bind(this)
+        ], [
+            global.display,
+            'window-entered-monitor',
+            this._updateLocationMap.bind(this)
+        ], [
+            global.display,
+            'window-left-monitor',
+            this._updateLocationMap.bind(this)
+        ]);
+    }
+
+    destroy() {
+        this._signalsHandler.destroy();
```

Comment:
Also add here a `this._monitor.run_dispose();` so that we are more explicit on destroying the underlying GObject.